src/Symfony/Component/HttpKernel/Controller/ArgumentResolver.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\Controller;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\ControllerMetadata\ArgumentMetadataFactoryInterface;
+
+/**
+ * Responsible for the resolving of arguments passed to an action.
+ *
+ * @author Iltar van der Berg <kjarli@gmail.com>
+ */
+final class ArgumentResolver implements ArgumentResolverInterface
+{
+    private $argumentMetadataFactory;
+
+    /**
+     * @var ArgumentValueResolverInterface[]
+     */
+    private $argumentValueResolvers;
+
+    public function __construct(ArgumentMetadataFactoryInterface $argumentMetadataFactory = null, array $argumentValueResolvers = array())
+    {
+        $this->argumentMetadataFactory = $argumentMetadataFactory;
+        $this->argumentValueResolvers = $argumentValueResolvers;
+    }
+
+    public function getArguments(Request $request, $controller)
+    {
+        $arguments = array();
+
+        foreach ($this->argumentMetadataFactory->createArgumentMetadata($controller) as $metadata) {
+            foreach ($this->argumentValueResolvers as $resolver) {
+                if (!$resolver->supports($request, $metadata)) {
+                    continue;
+                }
+
+                $resolved = $resolver->resolve($request, $metadata);
+
+                if (!$resolved instanceof \Generator) {
```

Comment:
@xabbuh in theory it shouldn't return anything else and a clear exception
is thrown when not the case. The decision for yield was that it's a simple
syntax and uniform for every case. Ideally I'd return the resolve value,
but that means that the variadic version breaks. If it's allowed to also
return iterators and such, it would make it too easy to break things. If
you would return 2 array items, it will become the wrong sequence of
arguments and break your method.

In my opinion having just yield is a good idea because it limits the usage
and reduces the chance of breaking. If you have an idea on how to approve
it, I'm all ears though

On Sun, Apr 10, 2016 at 1:41 PM Christian Flothmann <
notifications@github.com> wrote:

> In src/Symfony/Component/HttpKernel/Controller/ArgumentResolver.php
> <https://github.com/symfony/symfony/pull/18308#discussion_r59132159>:
>
> > +        $this->argumentValueResolvers = $argumentValueResolvers;
> > +    }
> > +
> > +    public function getArguments(Request $request, $controller)
> > +    {
> > +        $arguments = array();
> > +
> > +        foreach ($this->argumentMetadataFactory->createArgumentMetadata($controller) as $metadata) {
> > +            foreach ($this->argumentValueResolvers as $resolver) {
> > +                if (!$resolver->supports($request, $metadata)) {
> > +                    continue;
> > +                }
> > +
> > +                $resolved = $resolver->resolve($request, $metadata);
> > +
> > +                if (!$resolved instanceof \Generator) {
>
> @WouterJ <https://github.com/WouterJ> If we don't expect other code to
> make use of the return values, we should remove the
> ArgumentResolverInterface. Having such an interface calls for custom
> implementations imo which will then probably rely on the return type.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/symfony/symfony/pull/18308/files/b1e12875553883b0c3a01a7581109e0bc946a68a#r59132159>
>